Add support for async functions in funqy-knative

matejvasek opened the pull request.

matejvasek committed:
added support for async functions in funqy

user functions now may return CompletionStage

matejvasek commented:
@patriot1burke I don't know if you are working on async funs, if you do just close the PR.

matejvasek commented:
@patriot1burke I think this is ready for review.

patriot1burke commented on extensions/funqy/funqy-knative/deployment/src/test/java/io/quarkus/funqy/test/GreetingFunctions.java:
```diff
@@ -9,12 +15,25 @@
-        greeting.setMessage(message);
-        greeting.setName(name.getName());
-        return greeting;
+    public CompletionStage<Greeting> greet(Identity name) {
```
Do not delete the old synchronous function.  Write a new test for CompletionStage<> return value.  Also, please duplicate the separate CompletionStage test in Lambda and HTTP.

patriot1burke reviewed (changes requested):
Please add some processing to FunctionScannerBuildStep.scanFunctions to make sure that developer has declared a generic type for the CompletionStage.  If they haven't, throw a BuildException

matejvasek commented:
@patriot1burke I rebased and force pushed. I hope you don't mind. I added tests for both aync and async version as you requested. Also the build won't allow  unparameterized  CompletionStage.

matejvasek commented:
I could squash the commits If you want.

matejvasek commented:
So far vertx impls (knative and http) should by async now. With amazon lambda impl it is little different. It seems that main polling loop is blocking which means we must wait for CompletionStage  blockingly anyway. Maybe that could be changed see last commit: https://github.com/matejvasek/quarkus/tree/async-amazon-experiment.

patriot1burke commented:
The plan was to eventually rewrite the loop in vertx.  There is some context propagation that needs to be done though unfortunately.  See the xray extension which has a hack that does a code substitution to propagate a TraceId.  I don't know if other parts of the AWS stack make similar assumptions.

matejvasek commented:
Note: I really don't have knowledge how aws lambda run-time should work. Maybe the polling loop is supposed to be blocking.

patriot1burke commented:
I don't know either about whether the loop is supposed to be blocking.  Are you familiar with Quarkus deployment?  If you aren't I can either add the @BuildStep I talked about earlier, but if you want to learn, its a simple addition to get you to learn it.

matejvasek commented:
What buildsetp do you mean?

matejvasek commented:
> Are you familiar with Quarkus deployment?

I am not.

patriot1burke commented:
Just squash commits and I'll merge then.  I'll do the buildstep.

matejvasek commented:
> Please add some processing to FunctionScannerBuildStep.scanFunctions to make sure that developer has declared a generic type for the CompletionStage. If they haven't, throw a BuildException

Do you mean this? I am just throwing IllegalArgumentException in my code. When it bubbles up it interrupts build.

matejvasek commented:
I am going to squash code to one commit and force-push it. Ok?

matejvasek commented:
>Do you mean this? I am just throwing IllegalArgumentException in my code. When it bubbles up it interrupts build.

Ok. It doesn't. It crash at runtime.

matejvasek commented:
It crashes only native build.

matejvasek commented:
@patriot1burke I've just squashed it.

matejvasek committed:
forking for funqy amazon lambda integration tests

needed due static initialization
